src/bridge/cockpitdbusmachines.c

```diff
@@ -0,0 +1,451 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include <glob.h>
+#include <string.h>
+#include <errno.h>
+#include <json-glib/json-glib.h>
+
+#include "cockpitdbusinternal.h"
+#include "common/cockpitconf.h"
+#include "common/cockpitsystem.h"
+
+#define MACHINES_SIG "a{sa{sv}}"
+
+/* counts the number of file change events that have not yet gotten a PropertiesChanged signal*/
+static guint pending_updates;
+
+static const char*
+machines_json_dir (void)
+{
+  static gchar *path = NULL;
+  if (path == NULL)
+    {
+      const gchar * const* dirs = cockpit_conf_get_dirs ();
+
+      for (gint i = 0; dirs[i]; ++i)
+        {
+          g_free (path);
+          path = g_build_filename (dirs[i], "cockpit", "machines.d", NULL);
+          if (g_file_test (path, G_FILE_TEST_IS_DIR))
+            break;
+        }
+    }
+  return path;
+}
+
+static int
+glob_err_func (const char *epath, int eerrno)
+{
+  /* just log the error for debugging */
+  if (eerrno != ENOENT)
+    g_warning ("Cannot read %s: %s", epath, strerror (eerrno));
```

Comment:
Suggest using similar format to other warnings: ```g_warning ("%s: cannot read: %s", epath, g_strerror (eerno));```

In addition, note use of g_strerror().